File: javascript/crypto-square/crypto-square.js
const Crypto = function(input) {
  this.input = input;
};

Crypto.prototype.normalizePlaintext = function() {
  return this.input.toLowerCase().replace(/[^a-z0-9]/g, '');
};

Crypto.prototype.size = function() {
  return Math.ceil(Math.sqrt(this.normalizePlaintext(this.input).length));
};

Crypto.prototype.plaintextSegments = function() {
  let out    = [];
  let input  = this.normalizePlaintext(this.input);
  const size = this.size();

  while (input.length > 0) {
    out.push(input.slice(0, size));
    input = input.slice(size);
  }
  return out;
};

Crypto.prototype.ciphertext = function() {
  const segments = this.plaintextSegments();
  let out = "";

  for (var x=0; x<segments[0].length; x++) {
    for (var y=0; y<segments.length; y++) {
      if (segments[y][x] != undefined) {
        out += segments[y][x];
      }
    }
  }
  return out;
};

module.exports = Crypto;

File: ruby/crypto-square/crypto_square.rb
class Crypto
  def initialize(input)
    @input = input
  end

  def normalize_plaintext
    @input.downcase.gsub(/[^a-z0-9]/, '')
  end

  def size
    Math.sqrt(normalize_plaintext.length).ceil
  end

  def plaintext_segments
    normalize_plaintext.chars.each_slice(size).collect { |i| i.join }.to_a
  end

  def normalize_ciphertext
    transpose.join(' ')
  end

  def ciphertext
    transpose.join
  end

  private

  def transpose
    segs = plaintext_segments.map { |i| Array.new(size) { |j| i[j] or '' } }
    segs.transpose.map { |i| i.join }
  end
end
